flink-connectors/flink-connector-hbase-2.2/src/test/java/org/apache/flink/connector/hbase/HBaseConnectorITCase.java

```diff
@@ -0,0 +1,591 @@
+			((TableEnvironmentInternal) tEnv).registerTableSourceInternal("hTable", hbaseTable);
+		} else {
+			tEnv.executeSql(
+					"CREATE TABLE hTable (" +
+							" family1 ROW<col1 INT>," +
+							" family2 ROW<col1 STRING, col2 BIGINT>," +
+							" family3 ROW<col1 DOUBLE, col2 BOOLEAN, col3 STRING>," +
+							" rowkey INT," +
+							" PRIMARY KEY (rowkey) NOT ENFORCED" +
+							") WITH (" +
+							" 'connector' = 'hbase-2.2'," +
+							" 'table-name' = '" + TEST_TABLE_1 + "'," +
+							" 'zookeeper.quorum' = '" + getZookeeperQuorum() + "'" +
+					")");
+		}
+
+		Table table = tEnv.sqlQuery("SELECT " +
+				"  h.family1.col1, " +
+				"  h.family3.col1, " +
+				"  h.family3.col2, " +
+				"  h.family3.col3 " +
+				"FROM hTable AS h");
+
+		List<Row> results = CollectionUtil.iteratorToList(table.execute().collect());
+		String expected =
+				"10,1.01,false,Welt-1\n" +
+						"20,2.02,true,Welt-2\n" +
+						"30,3.03,false,Welt-3\n" +
+						"40,4.04,true,Welt-4\n" +
+						"50,5.05,false,Welt-5\n" +
+						"60,6.06,true,Welt-6\n" +
+						"70,7.07,false,Welt-7\n" +
+						"80,8.08,true,Welt-8\n";
+
+		TestBaseUtils.compareResultAsText(results, expected);
+	}
+
+	@Test
+	public void testTableSourceFieldOrder() {
+		TableEnvironment tEnv = createBatchTableEnv();
+
+		if (isLegacyConnector) {
+			HBaseTableSource hbaseTable = new HBaseTableSource(getConf(), TEST_TABLE_1);
+			// shuffle order of column registration
+			hbaseTable.setRowKey(ROW_KEY, Integer.class);
+			hbaseTable.addColumn(FAMILY2, F2COL1, String.class);
+			hbaseTable.addColumn(FAMILY3, F3COL1, Double.class);
+			hbaseTable.addColumn(FAMILY1, F1COL1, Integer.class);
+			hbaseTable.addColumn(FAMILY2, F2COL2, Long.class);
+			hbaseTable.addColumn(FAMILY3, F3COL2, Boolean.class);
+			hbaseTable.addColumn(FAMILY3, F3COL3, String.class);
+			((TableEnvironmentInternal) tEnv).registerTableSourceInternal("hTable", hbaseTable);
+		} else {
+			tEnv.executeSql(
+					"CREATE TABLE hTable (" +
+							" rowkey INT PRIMARY KEY," +
+							" family2 ROW<col1 STRING, col2 BIGINT>," +
+							" family3 ROW<col1 DOUBLE, col2 BOOLEAN, col3 STRING>," +
+							" family1 ROW<col1 INT>" +
+							") WITH (" +
+							" 'connector' = 'hbase-2.2'," +
+							" 'table-name' = '" + TEST_TABLE_1 + "'," +
+							" 'zookeeper.quorum' = '" + getZookeeperQuorum() + "'" +
+					")");
+		}
+
+		Table table = tEnv.sqlQuery("SELECT * FROM hTable AS h");
+
+		List<Row> results = CollectionUtil.iteratorToList(table.execute().collect());
+		String expected =
+				"1,Hello-1,100,1.01,false,Welt-1,10\n" +
+						"2,Hello-2,200,2.02,true,Welt-2,20\n" +
+						"3,Hello-3,300,3.03,false,Welt-3,30\n" +
+						"4,null,400,4.04,true,Welt-4,40\n" +
+						"5,Hello-5,500,5.05,false,Welt-5,50\n" +
+						"6,Hello-6,600,6.06,true,Welt-6,60\n" +
+						"7,Hello-7,700,7.07,false,Welt-7,70\n" +
+						"8,null,800,8.08,true,Welt-8,80\n";
+
+		TestBaseUtils.compareResultAsText(results, expected);
+	}
+
+	@Test
```

Comment:
Could we add  testTableSourceWithTableAPI here when rebase FLINK-18604?

Reply:
Made sure that all the modifications since the first patch are present at to both 1.4 and 2.2